go/vt/vtctld/vttablet_proxy.go

```diff
@@ -0,0 +1,81 @@
+/*
+Copyright 2020 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vtctld
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"strings"
+	"sync"
+
+	"vitess.io/vitess/go/vt/log"
+)
+
+var (
+	proxyTablets = flag.Bool("proxy_tablets", false, "Setting this true will make vtctld proxy the tablet status instead of redirecting to them")
+
+	proxyMu sync.Mutex
+	remotes = make(map[string]*httputil.ReverseProxy)
+)
+
+func addRemote(tabletID, path string) {
+	u, err := url.Parse(path)
+	if err != nil {
+		log.Errorf("Error parsing URL %v: %v", path, err)
+		return
+	}
+
+	proxyMu.Lock()
+	defer proxyMu.Unlock()
+	if _, ok := remotes[tabletID]; ok {
+		return
+	}
+
+	rp := &httputil.ReverseProxy{}
+	rp.Director = func(req *http.Request) {
+		splits := strings.SplitN(req.URL.Path, "/", 4)
+		if len(splits) < 4 {
+			return
+		}
+		req.URL.Scheme = u.Scheme
+		req.URL.Host = u.Host
+		req.URL.Path = "/" + splits[3]
+	}
+
+	prefixPath := fmt.Sprintf("/vttablet/%s/", tabletID)
+
+	rp.ModifyResponse = func(r *http.Response) error {
+		b, _ := ioutil.ReadAll(r.Body)
+		b = bytes.ReplaceAll(b, []byte(`href="/`), []byte(fmt.Sprintf(`href="%s`, prefixPath)))
+		b = bytes.ReplaceAll(b, []byte(`href=/`), []byte(fmt.Sprintf(`href=%s`, prefixPath)))
+		r.Body = ioutil.NopCloser(bytes.NewBuffer(b))
+		r.Header["Content-Length"] = []string{fmt.Sprint(len(b))}
+		// Don't forget redirects
+		loc := r.Header["Location"]
+		for i, v := range loc {
+			loc[i] = strings.Replace(v, "/", prefixPath, 1)
+		}
+		return nil
+	}
+	http.Handle(prefixPath, rp)
```

Comment:
It would be more natural to respond to tablets coming/going and changing addresses if there were only one handler, which looks at a cached lookup table of tablet aliases to addresses.

Reply:
> It would be more natural to respond to tablets coming/going and changing addresses if there were only one handler, which looks at a cached lookup table of tablet aliases to addresses.

This is what I started out with. The problem with the unified approach is that there is no easy way to associate a response to a request, which is needed for rewriting URLs to match the tablet id. I looked at making the vttablets send an extra header identifying the tablet id. But the one-proxy-per-tablet felt much simpler.

What I'll instead do is replace the per-tablet proxy every time. I looked at the code. It's just metdata (no goroutines etc). So, it should be fine.

Reply:
Oh I see. Couldn't you still have a single handler though? If reverse proxy objects are cheap to make, you can just make a new one for each request you handle and use it once. Then you don't have handlers slowly leaking in cases when tablet IDs go away.

Reply:
Currently, the page that renders the tablets is the one that causes the map entry to be created.

Thinking about it again, there's no real need for that page to be controlling the redirects. We could directly `GetTablet` at the time of request, create the redirect and forward it. Let me update the other PR to reflect this.